### PR TITLE
feat(zero-payload): Implement read and write for zero payload artifacts

### PR DESCRIPTION
**Pull request by MaciejTe**
feat(zero-payload): Implement read and write for zero payload artifacts

Changelog: Title
Ticket: MEN-2586
Signed-off-by: Maciej Tomczuk <maciej.tomczuk@northern.tech>

### CODE REVIEW

File: cli/cli_test.go [hunk 1]
```diff
@@ -157,3 +157,32 @@ func TestModuleImageWithoutPayload(t *testing.T) {
...
+		"bootstrap-artifact",
+		"-t", "dummy",
+		"-n", "dummy",
+		"-G", "dep_gr",
+		"-g", "pr_gr",
+		"--clears-provides", "cl_pr",
+		"-d", "dep:val",
+		"-p", "pr:val",
+		"-o", menderName,
+	})
```
Comment (kacf) [inline cli/cli_test.go:182]: AFAICS, we have nothing which tests the reading part. This is evident in that the [reader code is never hit](https://coveralls.io/jobs/102652485/source_files/23088432295#L687). We should use the `read` command at least once to verify that it works correctly for empty payloads.

OTOH the high level `read` command could be a bit clunky to use when checking the details, so I suppose there should be a few tests which calls the reader directly and verifies that the returned data is correct. I'll leave the details to you to decide what's most convenient.

File: cli/cli.go [hunk 2]
```diff
@@ -393,13 +393,61 @@ func getCliContext() *cli.App {
...
+		clearsArtifactProvides,
+		noDefaultClearsArtifactProvides,
+		payloadProvides,
+		payloadDepends,
+		/////////////////////////
+		// Version 3 specifics.//
+		/////////////////////////
+		artifactNameDepends,
+		artifactProvidesGroup,
+		artifactDependsGroups,
```
Comment (kacf) [inline cli/cli.go:438]: * The `--no-default` options should not be here (see also my other comment about default provides).
* `--key` argument needs to be here.

File: artifact/metadata.go [hunk 3]
```diff
@@ -74,7 +74,7 @@ func (i *Info) Write(p []byte) (n int, err error) {
 // UpdateType provides information about the type of update.
 // At the moment the only built-in type is "rootfs-image".
 type UpdateType struct {
-	Type string `json:"type"`
+	Type *string `json:"type"`
```
Comment (kacf) [inline artifact/metadata.go:77]: The changes to the artifact format need to be documented in `Documentation/artifact-format-v3.md`. Document that `type` can be `null`, and that this has the special properties:

* `data/xxxx.tar.gz` must be missing or empty.
* No `meta-data`.
* No augmented artifacts.

File: handlers/bootstrap_artifact.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+// Return type of original (non-augmented) update, if any.
+func (b *BootstrapArtifact) GetUpdateOriginalType() string {
+	return ""
+}
+
+func (b *BootstrapArtifact) GetUpdateDepends() (artifact.TypeInfoDepends, error) {
+	if b.typeInfoV3 == nil {
+		return nil, nil
+	}
+	return b.typeInfoV3.ArtifactDepends, nil
```
Comment (kacf) [inline handlers/bootstrap_artifact.go:58]: Same here, unconditionally return the bottom result?

**Review by kacf — CHANGES_REQUESTED** [review]
Phew! That's a quite a large one. Nice work here, I have many comments, but they are primarily about the details, the big picture here is indeed good.

File: artifact/metadata_test.go [hunk 5]
```diff
@@ -25,6 +25,10 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
+func UpdateTypePtr(s string) *string {
+	return &s
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply artifact/metadata_test.go:29]: Yes, exactly :D

File: cli/write.go [hunk 6]
```diff
@@ -133,6 +138,118 @@ func createRootfsFromSSH(c *cli.Context) (string, error) {
...
+		)
+	}
+
+	if err := validateInput(c); err != nil {
+		Log.Error(err.Error())
+		return err
+	}
+
+	// set the default name
+	name := "bootstrap.mender"
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply cli/write.go:173]: Both options are equally good in my opinion - on the one hand, `artifact.mender` is coherent with other writers, on the other hand, `bootstrap.mender` clearly shows there's no data and it's different type of artifact. I think I'll leave that call to you ;-)

**Comment by kjaskiewiczz** [comment]
@MaciejTe the PR is targeting feature branch, but yesterday I talked to @kacf, and we think it's ok to simply target master

File: areader/reader.go [hunk 7]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2021 Northern.tech AS
+// Copyright 2022 Northern.tech AS
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply areader/reader.go:1]: done!

File: cli/write.go [hunk 8]
```diff
@@ -133,6 +138,118 @@ func createRootfsFromSSH(c *cli.Context) (string, error) {
 	return rootfsFilename, nil
 }
 
+func makeEmptyUpdates(ctx *cli.Context) (*awriter.Updates, error) {
+	handler := handlers.NewBootstrapArtifact()
+
+	dataFiles := make([](*handlers.DataFile), 0, len(ctx.StringSlice("file")))
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply cli/write.go]: done!

File: cli/write.go [hunk 9]
```diff
@@ -133,6 +138,118 @@ func createRootfsFromSSH(c *cli.Context) (string, error) {
...
+		ArtifactName:  c.String("artifact-name"),
+		ArtifactGroup: c.String("provides-group"),
+	}
+
+	upd, err := makeEmptyUpdates(c)
+	if err != nil {
+		return err
+	}
+
+	typeInfoV3, _, err := makeTypeInfo(c)
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply cli/write.go:216]: Fixed; please let me know if current solution satisfies the requirements.

File: cli/write.go [hunk 10]
```diff
@@ -480,6 +603,9 @@ func applySoftwareVersionToTypeInfoProvides(
 	if ctx.Command.Name == "module-image" {
 		softwareNameDefault = ctx.String("type")
 	}
+	if ctx.Command.Name == "bootstrap-artifact" {
+		softwareNameDefault = "null"
+	}
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply cli/write.go:608]: done!

File: cli/write.go [hunk 11]
```diff
@@ -554,6 +680,8 @@ func makeClearsArtifactProvides(ctx *cli.Context) ([]string, error) {
 		}
 	} else if ctx.Command.Name == "module-image" {
 		softwareName = ctx.String("type") + "."
+	} else if ctx.Command.Name == "bootstrap-artifact" {
+		softwareName = "null" + "."
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply cli/write.go]: done!

File: handlers/bootstrap_artifact.go [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+	if err := writeTypeInfoV3(&WriteInfoArgs{
+		tarWriter:  args.TarWriter,
+		dir:        path,
+		typeinfov3: args.TypeInfoV3,
+	}); err != nil {
+		return errors.Wrap(err, "ComposeHeader: ")
+	}
+
+	if len(args.MetaData) > 0 {
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply handlers/bootstrap_artifact.go]: done!

File: areader/reader.go [hunk 13]
```diff
@@ -873,17 +876,20 @@ func (ar *Reader) readHeaderUpdate(tr *tar.Reader, hdr *tar.Header, augmented bo
 		// but they may exist if another tool was used to create the
 		// artifact.
 		if hdr.Typeflag != tar.TypeDir {
-			updNo, err := getUpdateNoFromHeaderPath(hdr.Name)
-			if err != nil {
-				return errors.Wrapf(err, "reader: error getting header Payload number")
-			}
+			// MEN-2586 - empty tar files are allowed in zero-payload artifacts
+			if hdr.Name != "" {
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply areader/reader.go]: It looks it's a development relic ;-) thanks for catching this!

File: handlers/bootstrap_artifact.go [hunk 14]
```diff
@@ -52,26 +52,38 @@ func (b *BootstrapArtifact) GetUpdateOriginalType() string {
...
 func (b *BootstrapArtifact) GetUpdateMetaData() (map[string]interface{}, error) {
 	return nil, nil
 }
+
 func (b *BootstrapArtifact) GetUpdateClearsProvides() []string {
-	return nil
+	if b.typeInfoV3 == nil {
+		return nil
+	}
+	return b.typeInfoV3.ClearsArtifactProvides
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply handlers/bootstrap_artifact.go]: done!

File: handlers/bootstrap_artifact.go [hunk 15]
```diff
@@ -131,6 +146,76 @@ func (b *BootstrapArtifact) ComposeHeader(args *ComposeHeaderArgs) error {
...
+func (b *BootstrapArtifact) ReadHeader(r io.Reader, path string, version int, augmented bool) error {
+	b.version = version
+	switch {
+	case filepath.Base(path) == "type-info":
+		dec := json.NewDecoder(r)
+		err := dec.Decode(&b.typeInfoV3)
+		if err != nil {
+			return errors.Wrap(err, "error reading type-info")
+		}
+	case filepath.Base(path) == "meta-data":
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply handlers/bootstrap_artifact.go]: done!

File: cli/write.go [hunk 16]
```diff
@@ -604,6 +604,9 @@ func applySoftwareVersionToTypeInfoProvides(
 		softwareNameDefault = ctx.String("type")
 	}
 	if ctx.Command.Name == "bootstrap-artifact" {
+		if len(result) > 0 {
+			return result
+		}
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply cli/write.go]: done!

File: handlers/bootstrap_artifact_test.go [hunk 17]
```diff
@@ -0,0 +1,427 @@
...
+
+			// err = r.ReadHeader(bytes.NewReader([]byte(test.data)), test.name, test.version, true)
+			// if test.shouldErrAugmented {
+			// 	require.Error(t, err)
+			// 	if test.errMsgAugmented != "" {
+			// 		require.Contains(t, errors.Cause(err).Error(), test.errMsgAugmented)
+			// 	}
+			// } else {
+			// 	require.NoError(t, err)
+			// }
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply handlers/bootstrap_artifact_test.go]: done!

File: cli/cli.go (hunk 2 above)
Comment (MaciejTe) [reply cli/cli.go]: I removed the `--no-default` options, but I'm not sure which `--key` argument needs to be added  - can you please specify  it exactly?

File: handlers/bootstrap_artifact.go (hunk 4 above)
Comment (MaciejTe) [reply handlers/bootstrap_artifact.go:55]: done!

File: cli/write.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (kacf) [reply cli/write.go:173]: Ok, if it's all up to me I think I vote for `artifact.mender`. It's true that it is a different type of artifact, but there are some niche uses of this type even outside of the bootstrap case (when users want to reset their database content), so we might as well keep the generic name.

File: Documentation/artifact-format-v3.md [hunk 18]
```diff
@@ -198,8 +198,11 @@ to several distinct components to be contained in one Artifact. For example,
 there may be an update to a file, and a package install contained in the same
 Artifact. For full rootfs updates, there will usually be only one payload.
 
-`type` is the type of payload contained within the image. At the moment there is
-one built-in type, `rootfs-image`, and all other strings will trigger use of
+`type` is the type of payload contained within the image. At the moment there are
+two built-in types:
+1. `rootfs-image`, 
+2. `null` (used in [empty payload artifacts](#Empty-payload-artifacts)),
```
Comment (kacf) [inline Documentation/artifact-format-v3.md:204]: Perhaps put double quotes around `"rootfs-image"` so it's clear that `null` is an actual `null` value, and not a string.

**Review by kacf — COMMENTED** [review]
Looks good, I have resolved everything which I agreed with. A couple of comments below, but it's pretty close now.

File: cli/cli_test.go (hunk 1 above)
Comment (MaciejTe) [reply cli/cli_test.go:182]: Done! Please take a look and see if current solution satisfies you.

Comment (kacf) [reply cli/cli_test.go:182]: 👍

File: cli/write.go [hunk 19]
```diff
@@ -554,6 +680,8 @@ func makeClearsArtifactProvides(ctx *cli.Context) ([]string, error) {
 		}
 	} else if ctx.Command.Name == "module-image" {
 		softwareName = ctx.String("type") + "."
+	} else if ctx.Command.Name == "bootstrap-artifact" {
+		return list, nil
```
[the remark this answers is not in the document]
Comment (MaciejTe) [reply cli/write.go]: done!

File: Documentation/artifact-format-v3.md (hunk 18 above)
Comment (MaciejTe) [reply Documentation/artifact-format-v3.md:204]: done!

File: cli/cli.go (hunk 2 above)
Comment (MaciejTe) [reply cli/cli.go]: Done (both things)!

Comment (kacf) [reply cli/cli.go]: Perfect, looks great!

**Comment by MaciejTe** [comment]
@mender-test-bot start pipeline

**Comment by kacf** [comment]
:checkered_flag: :grin:

**Comment by lluiscampos** [comment]
According to the [acceptance criteria](https://tracker.mender.io/browse/MEN-2582), this one should have gone into a feature branch. @MaciejTe @kacf Was it an oversight or a planned change? In other words, should we revert and submit to feature branch?

**Comment by kacf** [comment]
It was a deliberate choice. Since the empty payload feature of mender-artifact works independently of bootstrap artifact support in the client, we didn't see a reason to use a feature branch.

**Pull request merged**